Downloading metadata configurations from the server

Amejia481 opened the pull request:
close #1653

Amejia481 committed:
Config file parser #1653

-Adding Android and JVM test for MetaConfigurationImporter
-Adding Domain Classes for Questions and Options
-Adding Converters for Domain question to DB Questions
-Adding Converters for Domain option to DB option
-Creating a share source for Instrumentation test and JVM tes
-Adding base class for  for mockWebservice for Android JVM test

ifoche committed:
merged with development

ifoche commented on app/src/androidTestereferrals/java/org/eyeseetea/malariacare/MetadataConfigurationDBImporterShould.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    private void shouldNotBeAnyQuestionInTheDB() {
+        shouldBeInDB(0, 0, 0);
```
@Amejia481 just like we did for the 1.0 version, this test is modifying the DB, and should use an in memory DB. Please give a look to PR #1757 and check out there how we finally made to use an alternative "staging" variant for the tests that uses an in memory DB. Can you bring that implementation to this PR and check that this test doesn't modify the app DB? Thanks!

Amejia481 replied:
Sure!

ifoche commented on app/src/ereferrals/java/org/eyeseetea/malariacare/data/sync/importer/metadata/configuration/IMetadataConfigurationDataSource.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.List;
+
+public interface IMetadataConfigurationDataSource {
```
If we want to follow the same arch that we used for the rest of the app, I think this interface should be in the main, and in the remote package. Then, from inside the variant we can implement that interface.

Amejia481 replied:
Ok, I agree with moving the interface to main, but I am a little confuse with putting it on **data.remote** package because there's also a **data.sync.import** package where there's similar information.

For example the classes related to CVSImporter are in the same package **data.sync.import** but in the **ereferals** variant.

Could you help, please, to clarify in  which one should I put the interface?

ifoche commented on app/src/ereferrals/java/org/eyeseetea/malariacare/data/sync/importer/metadata/configuration/IMetadataConfigurationApi.java:
```diff
@@ -0,0 +1,13 @@
+import retrofit2.Call;
+import retrofit2.http.GET;
+
+public interface IMetadataConfigurationApi {
```
To follow the naming convention I think we should rename this to xxxApiClientRetrofit just like with SurveyApiClientRetrofit

ifoche commented on app/src/ereferrals/java/org/eyeseetea/malariacare/data/sync/importer/metadata/configuration/model/MetadataConfigurationsApi.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.List;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class MetadataConfigurationsApi {
```
Wouldn't it better to name this MetadataConfiguration? This is a model representing a MetadataConfiguration, and not an API, right?

Amejia481 replied:
I named it API because it represents a model from the API response, but I can change it as you suggest

ifoche commented on app/src/main/java/org/eyeseetea/malariacare/data/database/converts/OptionConverterFromDomainModelToDB.java:
```diff
@@ -0,0 +1,21 @@
+
+    @NotNull
+    @Override
+    public OptionDB convert(@NotNull Option domainModel) {
```
Looks like you're following the visitor pattern here to make the conversion. If this is the case, we already have an standard naming convention for this that you can use. Check what's under sync -> exporter -> xxxVisitor. I think all that has to do with the IConverter could be adapted to the same naming convention

Amejia481 replied:
:+1:

Amejia481 replied:
@ifoche there also a similar name convention **xxxMaper** in the package *data.mappers* which one should I use?

ifoche replied:
I've reviewed those mappers and I think your base implementation is closer to the visitors one, so I guess it would be much easier for everybody to understand it if you use the Visitor naming standard. However, maybe we can talk about which standard to use in our next meeting, so in the future, we're clear on which standard to follow

Amejia481 replied:
:+1:

ifoche commented on app/src/main/java/org/eyeseetea/malariacare/data/di/Injector.java:
```diff
@@ -0,0 +1,100 @@
+import okhttp3.OkHttpClient;
+import okhttp3.logging.HttpLoggingInterceptor;
+
+public class Injector {
```
I'm a bit concerned about introducing this "Injector", that's even in a "di" package as if we were trying to use it as a dependency injector. I'm not sure here and I would like to know @xurxodev's opinion. Could we just create factories (that's something we've already used in the app before)? At least, this element is something that we should discuss with @xurxodev, that for sure.

xurxodev replied:
@ifoche I have looked this class a while, really this class is not a injector is a factory. I think we should rename this class and its package

ifoche commented on app/src/main/java/org/eyeseetea/malariacare/data/sync/importer/poeditor/IPOEditorApi.java:
```diff
@@ -0,0 +1,24 @@
+import retrofit2.http.FormUrlEncoded;
+import retrofit2.http.POST;
+
+public interface IPOEditorApi {
```
Following the naming convention, should this be called xxxApiClientRetrofit?

ifoche commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Option.java:
```diff
@@ -0,0 +1,58 @@
+package org.eyeseetea.malariacare.domain.entity;
+
+
+public class Option {
```
This domain entity is anemic. This is something @xurxodev has already warned us about (http://xurxodev.com/clean-architecture-code-smells-parte-2/). Can you propose a way to provide this entity with some logic?

ifoche commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Question.java:
```diff
@@ -0,0 +1,71 @@
+
+import java.util.List;
+
+public class Question {
```
Same comment as before. This entity looks anemic

ifoche reviewed (changes requested):
@Amejia481 this is just a first round of requested changes. I think with this PR we may even need some intervention of @xurxodev to agree with some architectural-dependent decisions. Can you please give a look to the comments? Thanks!

Amejia481 committed:
Bringing InMemory DB for test from #1757 #1678

Amejia481 committed:
Renaming IMetadataConfigurationApi to IMetadataApiConfiguration

Renaming IPOEditorApi to IPOApiEditor

Moving IMetadataConfigurationDataSource.java

Renaming Injector to AppFactory

Amejia481 committed:
Renaming Injector to AppFactory #1678
Renaming IConverter to IConverterVisitor

Amejia481 committed:
Renaming MetadataConfigurationsApi to MetadataApiConfigurations  #1678

ifoche committed:
rename package androidTestereferrals to androidTestEreferrals to make staging variant work, and refactor visitor interface name

ifoche committed:
refactor parser test

ifoche committed:
refactor AppFactory into independent factories depending on their use

ifoche committed:
minor refactor

ifoche committed:
fix unit tests

ifoche committed:
merged with development

ifoche committed:
refactor visitors names

ifoche committed:
rename to IPOEditorApi

ifoche reviewed (approved):
thanks @Amejia481
